packages/viewer/src/components/viewer-transform-widget/viewer-transform-widget.tsx

```diff
@@ -0,0 +1,356 @@
+   */
+  @Prop({ mutable: true })
+  public position?: Vector3.Vector3;
+
+  /**
+   * The controller that is responsible for performing transforms.
+   */
+  @Prop({ mutable: true })
+  public controller?: TransformController;
+
+  /**
+   * The color of the translation arrow on the x-axis. Defaults to `#ea3324`.
+   */
+  @Prop()
+  public xArrowColor: Color.Color | string = '#ea3324';
+
+  /**
+   * The color of the translation arrow on the y-axis. Defaults to `#4faf32`.
+   */
+  @Prop()
+  public yArrowColor: Color.Color | string = '#4faf32';
+
+  /**
+   * The color of the translation arrow on the z-axis. Defaults to `#0000ff`.
+   */
+  @Prop()
+  public zArrowColor: Color.Color | string = '#0000ff';
+
+  /**
+   * The color override of the hovered component. Defaults to `#ffff00`.
+   */
+  @Prop()
+  public hoveredColor: Color.Color | string = '#ffff00';
+
+  /**
+   * The color to display when persistence of a transform is pending. Defaults to `#cccccc`.
+   */
+  @Prop()
+  public disabledColor: Color.Color | string = '#cccccc';
+
+  /**
+   * @internal
+   * @private
+   *
+   * Visible for testing.
+   */
+  @Prop({ mutable: true })
+  public hovered?: Mesh;
+
+  private currentPosition?: Vector3.Vector3;
+
+  private widget?: TransformWidget;
+  private dragging?: Mesh;
+  private lastWorldPosition?: Vector3.Vector3;
+
+  private canvasBounds?: DOMRect;
+  private canvasResizeObserver!: ResizeObserver;
+  private canvasRef?: HTMLCanvasElement;
+
+  private hoveredChangeDisposable?: Disposable;
+
+  protected componentDidLoad(): void {
+    window.addEventListener('pointermove', this.handleWindowPointerMove);
+
+    this.canvasResizeObserver = new ResizeObserver(this.handleCanvasResize);
+
+    if (this.canvasRef != null) {
+      this.canvasResizeObserver.observe(this.canvasRef);
+
+      this.setupTransformWidget();
+    }
+  }
+
+  protected disconnectedCallback(): void {
+    window.removeEventListener('pointermove', this.handleWindowPointerMove);
+
+    this.canvasResizeObserver.disconnect();
+
+    this.hoveredChangeDisposable?.dispose();
+    this.widget?.dispose();
+  }
+
+  @Watch('viewer')
+  protected handleViewerChanged(
+    newViewer?: HTMLVertexViewerElement,
+    oldViewer?: HTMLVertexViewerElement
+  ): void {
+    oldViewer?.removeEventListener('frameDrawn', this.handleViewerFrameDrawn);
+    oldViewer?.removeEventListener(
+      'dimensionschange',
+      this.handleViewerDimensionsChanged
+    );
+    newViewer?.addEventListener('frameDrawn', this.handleViewerFrameDrawn);
+    newViewer?.addEventListener(
+      'dimensionschange',
+      this.handleViewerDimensionsChanged
+    );
+
+    if (newViewer?.stream != null) {
+      this.controller?.dispose();
+      this.controller = new TransformController(newViewer.stream);
+    }
+
+    if (this.canvasRef != null && newViewer != null) {
+      this.canvasRef.width = newViewer.clientWidth;
```

Comment:
Can this code be moved to a function and reused in the `handleViewerDemensionsChanged`? I see there that we use the viewer.viewport, but in this case we use the clientWidth. Can we use the viewport in both instances?